timemachine/cpp/src/nonbonded_interaction_group.cu

```diff
@@ -0,0 +1,482 @@
+    gpuErrchk(cudaMemset(d_nblist_x_, 0, N_ * 3 * sizeof(*d_nblist_x_))); // set non-sensical positions
+    gpuErrchk(cudaMalloc(&d_nblist_box_, 3 * 3 * sizeof(*d_nblist_x_)));
+    gpuErrchk(cudaMemset(d_nblist_box_, 0, 3 * 3 * sizeof(*d_nblist_x_)));
+    gpuErrchk(cudaMalloc(&d_rebuild_nblist_, 1 * sizeof(*d_rebuild_nblist_)));
+    gpuErrchk(cudaMallocHost(&p_rebuild_nblist_, 1 * sizeof(*p_rebuild_nblist_)));
+
+    gpuErrchk(cudaMalloc(&d_sort_keys_in_, N_ * sizeof(d_sort_keys_in_)));
+    gpuErrchk(cudaMalloc(&d_sort_keys_out_, N_ * sizeof(d_sort_keys_out_)));
+    gpuErrchk(cudaMalloc(&d_sort_vals_in_, N_ * sizeof(d_sort_vals_in_)));
+
+    // initialize hilbert curve
+    std::vector<unsigned int> bin_to_idx(256 * 256 * 256);
+    for (int i = 0; i < 256; i++) {
+        for (int j = 0; j < 256; j++) {
+            for (int k = 0; k < 256; k++) {
+
+                bitmask_t hilbert_coords[3];
+                hilbert_coords[0] = i;
+                hilbert_coords[1] = j;
+                hilbert_coords[2] = k;
+
+                unsigned int bin = static_cast<unsigned int>(hilbert_c2i(3, 8, hilbert_coords));
+                bin_to_idx[i * 256 * 256 + j * 256 + k] = bin;
+            }
+        }
+    }
+
+    gpuErrchk(cudaMalloc(&d_bin_to_idx_, 256 * 256 * 256 * sizeof(*d_bin_to_idx_)));
+    gpuErrchk(
+        cudaMemcpy(d_bin_to_idx_, &bin_to_idx[0], 256 * 256 * 256 * sizeof(*d_bin_to_idx_), cudaMemcpyHostToDevice));
+
+    // estimate size needed to do radix sorting, this can use uninitialized data.
+    cub::DeviceRadixSort::SortPairs(
+        d_sort_storage_,
+        d_sort_storage_bytes_,
+        d_sort_keys_in_,
+        d_sort_keys_out_,
+        d_sort_vals_in_,
+        d_perm_,
+        std::max(NC_, NR_));
+
+    gpuErrchk(cudaPeekAtLastError());
+    gpuErrchk(cudaMalloc(&d_sort_storage_, d_sort_storage_bytes_));
+};
+
+template <typename RealType, bool Interpolated>
+NonbondedInteractionGroup<RealType, Interpolated>::~NonbondedInteractionGroup() {
+    gpuErrchk(cudaFree(d_col_atom_idxs_));
+    gpuErrchk(cudaFree(d_row_atom_idxs_));
+
+    gpuErrchk(cudaFree(d_lambda_plane_idxs_));
+    gpuErrchk(cudaFree(d_lambda_offset_idxs_));
+    gpuErrchk(cudaFree(d_du_dp_buffer_));
+    gpuErrchk(cudaFree(d_perm_)); // nullptr if we never built nblist
+
+    gpuErrchk(cudaFree(d_bin_to_idx_));
+    gpuErrchk(cudaFree(d_sorted_x_));
+
+    gpuErrchk(cudaFree(d_w_));
+    gpuErrchk(cudaFree(d_dw_dl_));
+    gpuErrchk(cudaFree(d_sorted_w_));
+    gpuErrchk(cudaFree(d_sorted_dw_dl_));
+    gpuErrchk(cudaFree(d_unsorted_p_));
+    gpuErrchk(cudaFree(d_sorted_p_));
+    gpuErrchk(cudaFree(d_unsorted_dp_dl_));
+    gpuErrchk(cudaFree(d_sorted_dp_dl_));
+    gpuErrchk(cudaFree(d_sorted_du_dx_));
+    gpuErrchk(cudaFree(d_sorted_du_dp_));
+
+    gpuErrchk(cudaFree(d_sort_keys_in_));
+    gpuErrchk(cudaFree(d_sort_keys_out_));
+    gpuErrchk(cudaFree(d_sort_vals_in_));
+    gpuErrchk(cudaFree(d_sort_storage_));
```

Comment:
Possible refinement: Could some of these repetitions be avoided using the `DeviceBuffer` added in https://github.com/proteneer/timemachine/pull/576?

Reply:
Agree, but was thinking to refactor this in a separate PR (I initially started to refactor this, but it involved more substantial changes than I had anticipated and this PR was getting big as-is). Tagging #577 to track.